src/gt4py/next/ffront/fbuiltins.py

```diff
@@ -37,103 +39,138 @@
+    elif t is Tuple:
+        return ts.TupleType
+    elif hasattr(t, "__origin__") and t.__origin__ is Union:
+        return tuple(_type_conversion_helper(e) for e in t.__args__)  # type: ignore[attr-defined]
+    else:
+        raise AssertionError("Illegal type encountered.")
 
-_reduction_like = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[ts.DeferredType(constraint=ts.FieldType)],
-        pos_or_kw_args={"axis": ts.DeferredType(constraint=ts.DimensionType)},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=ts.FieldType),
-    )
-)
 
-neighbor_sum = _reduction_like
-max_over = _reduction_like
-min_over = _reduction_like
-
-broadcast = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[
-            ts.DeferredType(constraint=(ts.FieldType, ts.ScalarType)),
-            ts.DeferredType(constraint=ts.TupleType),
-        ],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=ts.FieldType),
-    )
-)
+def _type_conversion(t):
+    return ts.DeferredType(constraint=_type_conversion_helper(t))
 
-where = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[
-            ts.DeferredType(constraint=ts.FieldType),
-            ts.DeferredType(constraint=(ts.FieldType, ts.ScalarType, ts.TupleType)),
-            ts.DeferredType(constraint=(ts.FieldType, ts.ScalarType, ts.TupleType)),
-        ],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=(ts.FieldType, ts.TupleType)),
-    )
-)
 
-astype = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[
-            ts.DeferredType(constraint=ts.FieldType),
-            ts.DeferredType(constraint=ts.FunctionType),
-        ],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=ts.FieldType),
-    )
-)
+@dataclasses.dataclass(frozen=True)
+class BuiltInFunction(Generic[R, P]):
+    name: str = dataclasses.field(init=False)
+    # `function` can be used to provide a default implementation for all `Field` implementations,
+    # e.g. a fused multiply add could have a default implementation as a*b+c, but an optimized implementation for a specific `Field`
+    function: Callable[P, R]
 
-_unary_math_builtin = BuiltInFunction(
-    ts.FunctionType(
-        pos_only_args=[ts.DeferredType(constraint=(ts.ScalarType, ts.FieldType))],
-        pos_or_kw_args={},
-        kw_only_args={},
-        returns=ts.DeferredType(constraint=(ts.ScalarType, ts.FieldType)),
-    )
-)
+    def __post_init__(self):
+        # functools.update_wrapper(
+        #     self, self.function
+        # )  # TODO figure out keeping function annotations in autocomplete
+        object.__setattr__(self, "name", f"{self.function.__module__}.{self.function.__name__}")
+
+    def __call__(self, *args: Value, **options: Any) -> Value | tuple[Value, ...]:
```

Comment:
```suggestion
    def __call__(self, *args: Value, **options: Any) -> R:
```
and check the other type annotations